opensearch/setup_models.py

```diff
@@ -0,0 +1,562 @@
+    """
+    create an HTTP connector. handle any errors appropriately
+    """
+    if "name" not in connector_body:
+        die(f"missing 'name' field in connector body: {connector_body}")
+    id_maybe = connector_exists(connector_body["name"])
+    if id_maybe is not None:
+        return id_maybe
+    for i in range(attempts):
+        connector_response = opensearch_request(
+            "/_plugins/_ml/connectors/_create", connector_body, "POST", log_name=f"create_{connector_name}"
+        )
+        wait_time, successful, connector_id = handle_connector_response(connector_response)
+        if successful:
+            return connector_id
+        time.sleep(wait_time)
+    die(f"Failed to create connector after {attempts} attempts")
+
+
+def handle_connector_response(connector_response):
+    """
+    handle a create connector response, including errors
+    """
+    if "connector_id" in connector_response:
+        return 1, True, connector_response["connector_id"]
+    die(f"Error creating a connector: {connector_response}")
+
+
+def cycle_task(start_task, get_action, task_id=None, timeout=60):
+    """
+    Spawn a task and watch it until it completes, handling errors and retrying
+    - start_task: lambda that spawns a task or gets the running task (returns new task id)
+    - get_action: lambda that decides what to do based on task status (return value, wait time, try again)
+    - task_id: task id to watch
+    """
+    if task_id is None:
+        task_id = start_task()
+    deadline = time.time() + timeout
+    while time.time() < deadline:
+        return_value, wait_time, should_try_again = get_action(task_id)
+        if return_value is not None:
+            return return_value
+        if should_try_again:
+            task_id = start_task()
+        time.sleep(wait_time)
+    die(f"Could not complete task in {timeout} seconds")
+
+
+def construct_get_action_fn(handle_error_fn, is_complete_fn=lambda x: False):
+    """
+    Construct the lambda to hand to `cycle_task` as `get_action`
+    """
+
+    def inner_get_action(task_id):
+        task_status = opensearch_request(f"/_plugins/_ml/tasks/{task_id}", log_name=f"get_task_{task_id}")
+        if "state" not in task_status:
+            die(f"task status missing 'state' field: {task_status}")
+        state = task_status["state"]
+        if state in ["RUNNING", "CREATED"]:
+            # don't return anything, wait a sec, don't try again
+            return None, 1, False
+        elif state in ["COMPLETED"] or is_complete_fn():
+            # return model id, no waiting, don't try again
+            return task_status["model_id"], 0, False
+        elif state in ["FAILED"]:
+            # defer to error handling function
+            return handle_error_fn(task_status)
+        else:
+            die(f"Unrecognized task status: {task_status}")
+
+    return inner_get_action
+
+
+def construct_error_handler_fn():
```

Comment:
I don't think you need the wrapper function.
def foo(x):
    print(x)

def bar(f, x):
    f(x)

bar(foo, 3)
bar(foo, 7)
does what you expect.

You'd need it if you were capturing something, but you're not.